core/src/main/java/io/questdb/cairo/EngineMigration.java

```diff
@@ -0,0 +1,391 @@
+                                                // to meta file
+                                                // This is critical, table potentially left in unusable state
+                                                throw CairoException.instance(ff.errno())
+                                                        .put("failed to write updated version to table Metadata file [path=")
+                                                        .put(path.trimTo(plen))
+                                                        .put(",latestVersion=")
+                                                        .put(i);
+                                            }
+                                        }
+                                    }
+                                    return;
+                                }
+                                updateSuccess = false;
+                                throw CairoException.instance(ff.errno()).put("Could not update table [path=").put(path).put(']');
+                            } finally {
+                                ff.close(fd);
+                                path.trimTo(plen);
+                            }
+                        }
+                    }
+                }
+            });
+            LOG.info().$("upgraded tables to ").$(latestVersion).$();
+        }
+        return updateSuccess;
+    }
+
+    @FunctionalInterface
+    interface MigrationAction {
+        void migrate(MigrationContext context);
+    }
+
+    private static class MigrationActions {
+        private static void assignTableId(MigrationContext migrationContext) {
+            long mem = migrationContext.getTempMemory(8);
+            FilesFacade ff = migrationContext.getFf();
+            Path path = migrationContext.getTablePath();
+
+            LOG.info().$("setting table id in [path=").$(path).I$();
+            Unsafe.getUnsafe().putInt(mem, migrationContext.getNextTableId());
+            if (ff.write(migrationContext.getMetadataFd(), mem, Integer.BYTES, META_OFFSET_TABLE_ID) == Integer.BYTES) {
+                return;
+            }
+            throw CairoException.instance(ff.errno()).put("Could not update table id [path=").put(path).put(']');
+        }
+
+        private static void rebuildTransactionFile(MigrationContext migrationContext) {
+            // Update transaction file
+            // Before there was 1 int per symbol and list of removed partitions
+            // Now there is 2 ints per symbol and 4 longs per each non-removed partition
+
+            var path = migrationContext.getTablePath();
+            final var ff = migrationContext.getFf();
+            int pathDirLen = path.length();
+
+            path.concat(TXN_FILE_NAME).$();
+            var txMem = migrationContext.getRwMemory();
+            LOG.debug().$("opening for rw [path=").$(path).I$();
+            txMem.of(ff, path.$(), ff.getPageSize());
```

Comment:
All closed by the caller, including txMem (MigrationContext.rwMemory).
TxMem also closed by the callee only because there is no other method to release the files switched to in of method.